### ISSUE TITLE
Bug: [no-unused-expressions, no-empty-functions, possibly others...] extension rules crash with eslint v9.15.0

### ISSUE THREAD

**Issue by ext4cats** [typescript-eslint/typescript-eslint#10338]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

eslint crashes when using the latest version of node 22 and the latest version of typescript-eslint.

```
Oops! Something went wrong! :(

ESLint: 9.15.0

TypeError: Error while loading rule '@typescript-eslint/no-unused-expressions': Cannot read properties of undefined (reading 'allowShortCircuit')
Occurred while linting /home/ext4cats/Projects/tseslint-8.14.0-bug/index.js
    at Object.create (/home/ext4cats/Projects/tseslint-8.14.0-bug/node_modules/.pnpm/eslint@9.15.0/node_modules/eslint/lib/rules/no-unused-expressions.js:75:13)
    at create (/home/ext4cats/Projects/tseslint-8.14.0-bug/node_modules/.pnpm/@typescript-eslint+eslint-plugin@8.14.0_@typescript-eslint+parser@8.14.0_eslint@9.15.0_typesc_u2crwq53uno6k5fweinf7nowby/node_modules/@typescript-eslint/eslint-plugin/dist/rules/no-unused-expressions.js:28:32)
    at Object.create (/home/ext4cats/Projects/tseslint-8.14.0-bug/node_modules/.pnpm/@typescript-eslint+utils@8.14.0_eslint@9.15.0_typescript@5.6.3/node_modules/@typescript-eslint/utils/dist/eslint-utils/RuleCreator.js:31:20)
    at createRuleListeners (/home/ext4cats/Projects/tseslint-8.14.0-bug/node_modules/.pnpm/eslint@9.15.0/node_modules/eslint/lib/linter/linter.js:944:21)
    at /home/ext4cats/Projects/tseslint-8.14.0-bug/node_modules/.pnpm/eslint@9.15.0/node_modules/eslint/lib/linter/linter.js:1082:84
    at Array.forEach (<anonymous>)
    at runRules (/home/ext4cats/Projects/tseslint-8.14.0-bug/node_modules/.pnpm/eslint@9.15.0/node_modules/eslint/lib/linter/linter.js:1013:34)
    at #flatVerifyWithoutProcessors (/home/ext4cats/Projects/tseslint-8.14.0-bug/node_modules/.pnpm/eslint@9.15.0/node_modules/eslint/lib/linter/linter.js:1911:31)
    at Linter._verifyWithFlatConfigArrayAndWithoutProcessors (/home/ext4cats/Projects/tseslint-8.14.0-bug/node_modules/.pnpm/eslint@9.15.0/node_modules/eslint/lib/linter/linter.js:1993:49)
    at Linter._verifyWithFlatConfigArray (/home/ext4cats/Projects/tseslint-8.14.0-bug/node_modules/.pnpm/eslint@9.15.0/node_modules/eslint/lib/linter/linter.js:2082:21)
```

### Reproduction Repository Link

https://github.com/ext4cats/tseslint-8.14.0-bug

### Repro Steps

1) clone the repo
2) follow steps in readme

### Versions

| package                                |  version  |
| -------------------------------------- | --------- |
| `typescript-eslint`                    | `8.14.0`  |
| `TypeScript`                           | `5.6.3`   |
| `ESLint`                               | `9.15.0`  |
| `node`                                 | `22.11.0` |

**Comment by glass-ships** [comment]
Related to https://github.com/eslint/eslint/issues/19134

**Comment by ext4cats** [comment]
I downgraded my project to eslint v9.14.0 and the crash is gone. Seems to be a regression in eslint itself.

**Comment by AviVahl** [comment]
It looks like `@typescript-eslint/no-unused-expressions` uses eslint's original `no-unused-expressions`, calling its create function, passing in the context. `no-unused-expressions` appears to be expecting context.options to be defined (as an array of options), as explained here: https://eslint.org/docs/latest/extend/custom-rules#the-context-object

![image](https://github.com/user-attachments/assets/4b624aa9-285d-4411-8b36-203e989d7795)

for some reason, `@typescript-eslint/no-unused-expressions`'s create function appears to be receiving these options as a second parameter:

![image](https://github.com/user-attachments/assets/2656fac7-ce7b-4e8d-9605-46a255ca2068)

not sure if this inconsistency was always there, but typescript-eslint may want to propagate options in?

**Comment by AviVahl** [comment]
the crash is due to this change https://github.com/eslint/eslint/pull/17656/files#diff-b65ad05fc10a27a62ba130f4ec20fce33dfdfed6917e58581c91d59067aeb30b

which makes typescript-eslint's way of using it break.
the fix should _probably_ come in typescript-eslint, to match the rule's new context expectations.

**Comment by kirkwaiblinger** [comment]
Yes, as several have noted this is due to the `@typescript-eslint/no-unused-expressions` extension rule using brittle monkey-patching to compose our logic with that of the the base rule. When the base rule changes, so must our monkey-patch! I note that https://github.com/eslint/eslint/pull/17656 seems to much broader than just affecting `no-unused-expressions`, so I honestly wouldn't be surprised if other rules are broken with eslint v9.15.0 too for the same reason. Worth looking into.

**Comment by AviVahl** [comment]
search for `baseRule.create` in the rules directory. I assume all these could break in theory. not sure which ones are included in the recommended preset.

**Comment by AviVahl** [comment]
perhaps the fix should be to match the new upstream defaultOptions behavior, and even use it for typescript-eslint rules via context.options (instead of second parameter)?

**Comment by kirkwaiblinger** [comment]
> perhaps the fix should be to match the new upstream defaultOptions behavior, and even use it for typescript-eslint rules via context.options (instead of second parameter)?

The thing to keep in mind is that we'll want to stay backwards compatible with eslint versions before 9.15. See also https://github.com/typescript-eslint/typescript-eslint/issues/6456. 

If you end up poking around with this and find a good solution, do feel free to [raise a PR](https://typescript-eslint.io/contributing/pull-requests) 🙂

**Comment by AviVahl** [comment]
Considered it, but it seems like a wider change, so I'm hesitant given my limited knowledge of this codebase...

**Comment by kirkwaiblinger** [comment]
> Considered it, but it seems like a wider change, so I'm hesitant given my limited knowledge of this codebase...

Totally fair! Rest assured that the root cause of this type of bug is very well understood by each of the maintenance team members, and will be addressed as soon as our bandwidth permits, unless a community champion is able to get to it first 🙂

**Comment by bradzacher** [comment]
> for some reason, @typescript-eslint/no-unused-expressions's create function appears to be receiving these options as a second parameter:
> not sure if this inconsistency was always there...

@AviVahl our utilities have _always_ supported a top-level `defaultOptions` property for as long as they have existed!
Our API design was to pass the merged options as a second parameter to the `create` function rather than changing the options on the `context` object because it allowed greater flexibility -- it gives you the option to see BOTH the user's actual config _and_ the merged config -- which can sometimes be useful if you want to act differently when a user skipped an option but also avoid having `undefined`s around the place.

Unfortunately ESLint core chose a different location for the `defaultOptions` property (in `meta.defaultOptions` rather than as a top-level property) and they chose to overwrite the `context.options` rather than pass it as a separate argument.

In the next major we'll likely converge on their solution to deduplicate things.

**Comment by davidcarboni** [comment]
In case it's helpful: I'm seeing this on `@typescript-eslint/no-unused-expressions` (also fixed by downgrading eslint to ~9.14.0)

**Comment by Jason3S** [comment]
@bradzacher,

The issue is not limited to `@typescript-eslint`, `unicorn` also has issues.

```
TypeError: Error while loading rule 'unicorn/expiring-todo-comments': Cannot read properties of undefined (reading 'decoration')
```

Is the problem an issue with `defaultOptions` not getting correctly propagated or with code that cannot handle undefined options:

https://github.com/eslint/eslint/pull/17656/files#diff-295e5aae69e9308d6777e4522ca3337e15b8eaa604beaf30c679b45dd33cd20eR73-R79

The following will give a `TypeError: Cannot read properties of undefined ...`
if `context.options[0] === undefined`.
```
const [{
            allow,
            ignoreDestructuring,
            ignoreGlobals,
            ignoreImports,
            properties
        }] = context.options;
```

**Comment by ext4cats** [comment]
So, is all of this a regression in `eslint` or is it a case of `typescript-eslint` (and `unicorn` too apparently) exploiting behavior that's not part of the public plugin API? I can't tell because I'm not at all familiar with how `eslint` or its plugins work.

**Comment by ext4cats** [comment]
@glass-ships Smells like a breaking change sneaking its way into a minor update. I've already locked all of my projects to v9.14.0 so I should be fine for now. I might look into it and see if I can craft a fix this weekend.

If I can figure out the codebase in just a weekend, that is... I don't have high hopes, but I feel invested in this considering how I found it just an hour after v9.15.0 came out, after doing my routine `pnpm update` 😬

**Comment by glass-ships** [comment]
Agreed! If you do happen to give this a shot, feel free to link me to your PR, I'd be happy to contribute however I can.

**Comment by Jason3S** [comment]
This is the eslint patch I used to get it working with typescript-eslint and unicorn

```diff
diff --git a/lib/rules/no-unused-expressions.js b/lib/rules/no-unused-expressions.js
index fd1437c1606a56cbce147d040d601d0acef3d24a..1ac2d6e45fbb04f93d9cccdc5265bb38c3701aca 100644
--- a/lib/rules/no-unused-expressions.js
+++ b/lib/rules/no-unused-expressions.js
@@ -76,7 +76,7 @@ module.exports = {
             allowTernary,
             allowTaggedTemplates,
             enforceForJSX
-        }] = context.options;
+        } = {}] = context.options;
 
         /**
          * Has AST suggesting a directive.
diff --git a/lib/rules/no-warning-comments.js b/lib/rules/no-warning-comments.js
index 628f5a2ac513303cc3fe2cc9adcbf1ae54128747..a42e8318e028c6a3717afceef3c949bf8eab5827 100644
--- a/lib/rules/no-warning-comments.js
+++ b/lib/rules/no-warning-comments.js
@@ -64,7 +64,7 @@ module.exports = {
 
     create(context) {
         const sourceCode = context.sourceCode;
-        const [{ decoration, location, terms: warningTerms }] = context.options;
+        const [{ decoration, location, terms: warningTerms } = this.meta.defaultOptions[0]] = context.options;
         const escapedDecoration = escapeRegExp(decoration ? decoration.join("") : "");
         const selfConfigRegEx = /\bno-warning-comments\b/u;
```

**Comment by TheGreatJordach** [comment]
I had the same problem with V10.8.2 . it stopped happening if I downgraded the version to v9.14.0.

**Comment by bradzacher** [comment]
The issue is known. The fix is in progress.
Locking to avoid further noise.

Please subscribe to this issue to keep updated.

**Comment by bradzacher** [comment]
The fix has merged and will soon be released to our `canary` tag.
The fix will release to `latest` with our next release in a few days.
https://typescript-eslint.io/users/releases

**Comment by miraclemenikelechi** [comment]
> This is the eslint patch I used to get it working with typescript-eslint and unicorn
> 
> ```diff
> diff --git a/lib/rules/no-unused-expressions.js b/lib/rules/no-unused-expressions.js
> index fd1437c1606a56cbce147d040d601d0acef3d24a..1ac2d6e45fbb04f93d9cccdc5265bb38c3701aca 100644
> --- a/lib/rules/no-unused-expressions.js
> +++ b/lib/rules/no-unused-expressions.js
> @@ -76,7 +76,7 @@ module.exports = {
>              allowTernary,
>              allowTaggedTemplates,
>              enforceForJSX
> -        }] = context.options;
> +        } = {}] = context.options;
>  
>          /**
>           * Has AST suggesting a directive.
> diff --git a/lib/rules/no-warning-comments.js b/lib/rules/no-warning-comments.js
> index 628f5a2ac513303cc3fe2cc9adcbf1ae54128747..a42e8318e028c6a3717afceef3c949bf8eab5827 100644
> --- a/lib/rules/no-warning-comments.js
> +++ b/lib/rules/no-warning-comments.js
> @@ -64,7 +64,7 @@ module.exports = {
>  
>      create(context) {
>          const sourceCode = context.sourceCode;
> -        const [{ decoration, location, terms: warningTerms }] = context.options;
> +        const [{ decoration, location, terms: warningTerms } = this.meta.defaultOptions[0]] = context.options;
>          const escapedDecoration = escapeRegExp(decoration ? decoration.join("") : "");
>          const selfConfigRegEx = /\bno-warning-comments\b/u;
> ```

this works for now! thank you!

**Comment by SDAdham** [comment]
Hello, I just installed `canary` and the error is still there.

**Comment by ext4cats** [comment]
> Hello, I just installed `canary` and the error is still there.

Can you provide details and/or a reproduction?

**Comment by bradzacher** [comment]
If you think this isn't fixed, please file a new issue and provide a complete reproduction and we can investigate.

**Comment by kirkwaiblinger** [comment]
Update: The fix is now live in [our 8.15.0 release](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v8.15.0)! 🎉 

If you encounter issues when you try it out, please double-check that _each_ of the `@typescript-eslint/*` packages you're using is up-to-date (the most relevant one is `@typescript-eslint/eslint-plugin`), and create a minimal reproduction of the issue, before filing a new issue.
